nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/SwappablePriorityQueue.java

```diff
@@ -920,7 +921,14 @@ public SwapSummary recoverSwappedFlowFiles() {
         return new StandardSwapSummary(new QueueSize(swapFlowFileCount, swapByteCount), maxId, resourceClaims);
     }
 
+    public long getMaxActiveQueuedDuration(long fromTimestamp) {
+        // We want the oldest timestamp, which will be the min
+        return fromTimestamp - activeQueue.parallelStream().map(FlowFile::getLastQueueDate).filter(Objects::nonNull).min(Long::compareTo).orElse(fromTimestamp);
```

Comment:
Unfortunately, these numbers are not accurate if any FlowFiles have been swapped out. To get the max active queued duration, we would have to also consider any data in swap files. To do that efficiently, before swapping data out, the oldest timestamp would need to be written to the swap file header. Even scanning the header, though, would be too expensive to invoke regularly, so it would be important that this class also keep a mapping of swap file to earliest timestamp, and recover that on startup when recovering swapped data.

Additionally, we cannot access the `activeQueue` outside of a Read/Write lock, so that would need to be added in here for appropriate thread safety. Would also say that we should avoid any parallel streams here. The act of creating the parallel stream is almost guaranteed to be far more expensive than the cost of just iterating over all of the elements in the queue using a for-each.

Reply:
Understood and agreed. I will update accordingly.

Reply:
@markap14 Do you have a recommendation as to how to persist & retrieve the map of swapped queueDates or how to recreate it on startup?

Reply:
Disregard my previous question, I misunderstood how swap files worked. I now understand that all flowfiles are swapped in from a given swap file at the same time vs one at a time. Therefore we only need to know what the min last queued date is, as well as the total of all last queued dates and total number of swapped flow files. From that we can extrapolate anything else we may need.